File: w3/dz_java/src/Gun/Gun.java
package Gun;

public class Gun {
    private int amount;
    private int capacity;
    private boolean isReady;
    private String model;
    private int totalShots;

    public Gun() {
        this.amount = 0;
        this.capacity = 8;
        this.isReady = false;
        this.model = "Beretta";
        this.totalShots = 0;
    }

    public Gun(String model, int capacity) {
        this.amount = 0;
        this.capacity = capacity;
        this.isReady = false;
        this.model = model;
        this.totalShots = 0;
    }

    public int getAmount() {
        return amount;
    }

    public int getCapacity() {
        return capacity;
    }

    public boolean ready() {
        return isReady;
    }

    public String getModel() {
        return model;
    }

    public int getTotalShots() {
        return totalShots;
    }

    public void prepare() {
        if ( !this.isReady ) {
            this.isReady = true;
        }
    }

    public void reload() {
        this.amount = this.capacity;
        System.out.println("Gun reloaded!");
    }

    public void shoot() throws NotReady, OutOfRounds {
        if ( !this.isReady ) {
            throw new NotReady();
        }

        if ( this.amount == 0 ) {
            throw new OutOfRounds();
        }

        System.out.println("Bang!");
        this.amount -= 1;
        this.totalShots += 1;
    }

    public String toString() {
        return String.format("Gun model: " + this.getModel() + "\n"
                            + "Rounds/Capacity: " + this.getAmount() + "/" + this.getCapacity() + "\n"
                            + "Total shots: " + this.getTotalShots() + "\n"
                            + "Ready gun: " + this.ready());
    }
}

File: w3/dz_cpp/bcw4/bcw4/Gun/Gun.cpp
#include <iostream>
#include "Gun.h"

Gun::Gun(const std::string& model, int capacity) {
    this->amount = 0;
    this->capacity = capacity;
    this->isReady = false;
    this->model = model;
    this->totalShots = 0;
}

Gun::~Gun() {}

int Gun::getAmount() const {
    return this->amount;
}

int Gun::getCapacity() const {
    return this->capacity;
}

bool Gun::ready() const {
    return this->isReady;
}

const std::string& Gun::getModel() const {
    return this->model;
}

int Gun::getTotalShots() const {
    return this->totalShots;
}

void Gun::prepare() {
    if ( !this->isReady )
        this->isReady = true;
}

void Gun::reload() {
    this->amount = this->capacity;
    std::cout << "Gun reloaded!" <<std::endl;
}

void Gun::shoot() {
    if ( !this->ready() ) {
        throw NotReady();
    }

    if ( this->amount == 0 ) {
        throw OutOfRounds();
    }
    std::cout <<"Bang!" << std::endl;
    this->amount -= 1;
    this->totalShots += 1;
}

std::ostream& operator<<(std::ostream& out, const Gun& gun) {
    out << "Gun model: " << gun.getModel() << "\n";
    out << "rounds/capacity: " << gun.getAmount() << "/" << gun.getCapacity() << "\n";
    out << "Total shots: " << gun.getTotalShots() << "\n";
    out << "Ready gun: " << gun.ready() << std::endl;
    return out;
}